src/native/checksums.c

```diff
@@ -9,19 +9,29 @@
 #include "crt.h"
 #include "java_class_ids.h"
 
+/**
+ * Note: we use critical mem access in below functions to speed up checksums.
+ * This approach is the same as what OpenJDK uses for their checksum implementation.
+ * Think twice before using similar approach elsewhere as it can lead to stalling.
+ */
+
 jint crc32_common(
     JNIEnv *env,
     jbyteArray input,
     jint previous,
     const size_t start,
     size_t length,
     uint32_t (*checksum_fn)(const uint8_t *, size_t, uint32_t)) {
-    struct aws_byte_cursor c_byte_array = aws_jni_byte_cursor_from_jbyteArray_acquire(env, input);
+    struct aws_byte_cursor c_byte_array = aws_jni_byte_cursor_from_jbyteArray_critical_acquire(env, input);
+    if (AWS_UNLIKELY(c_byte_array.ptr == NULL)) {
+        return previous;
+    }
```

Comment:
Why are we returning previous here?

Reply:
we need to return something. what does not really matter as java should find exception on the stack